File: cpp/13-zmq_PubCStdin-Subjava/subscriber.cpp
#include <zmq.hpp>
#include <iostream>
#include <sstream>

char none[0];
int main (int argc, char *argv[]) {
	zmq::context_t context (1);
	zmq::socket_t subscriber (context, ZMQ_SUB);
	subscriber.connect("tcp://localhost:5556");
	subscriber.setsockopt(ZMQ_SUBSCRIBE, none);

	for (int count=0; count<4; count++) {
		char json[255];
		zmq::message_t update;

		subscriber.recv(&update);
		std::istringstream iss(static_cast<char*>(update.data()));
		iss >> json;

		std::cerr << "SUB: Received message [" << count<< "]: "
			  << json << "\n" << std::endl;
	}
	return 0;
}

File: cpp/13-zmq_PubCStdin-Subjava/Subscriber.java
import java.util.StringTokenizer;
import org.zeromq.ZMQ;

public class Subscriber {
	public static void main(String[] args) {
		ZMQ.Context context = ZMQ.context(1);
		ZMQ.Socket subscriber = context.socket(ZMQ.SUB);
		subscriber.connect("tcp://localhost:5556");
		subscriber.subscribe("");

		int count;
		for (count = 0; count < 5; count++) {
			String string = subscriber.recvStr(0).trim();
			StringTokenizer sscanf = new StringTokenizer(string, " ");
			String json = sscanf.nextToken();
			System.out.println("SUB: Got message [" + count + "]: " + json);
		}
		subscriber.close();
		context.term();
	}
}
